pkg/cdi/cache_test.go

```diff
@@ -812,6 +814,102 @@ devices:
 	}
 }
 
+func TestTooManyOpenFiles(t *testing.T) {
+	if runtime.GOOS != "linux" {
+		t.Skip("Test not applicable on Windows")
+	}
+
+	em, err := triggerEmfile()
+	require.NoError(t, err)
+	require.NotNil(t, em)
+	defer func() {
+		require.NoError(t, em.undo())
+	}()
+
+	_, err = syscall.Socket(syscall.AF_INET, syscall.SOCK_DGRAM, 0)
+	require.Equal(t, syscall.EMFILE, err)
+
+	cache := newCache(
+		WithAutoRefresh(true),
+	)
+	require.NotNil(t, cache)
+
+	// try to trigger original crash with a nil fsnotify.Watcher
+	_, _ = cache.InjectDevices(&oci.Spec{}, "vendor1.com/device=dev1")
+}
+
+type emfile struct {
+	limit syscall.Rlimit
+	fds   []int
+}
+
+func getProcStatusFdSize() (uint64, error) {
+	status, err := os.ReadFile("/proc/self/status")
+	if err != nil {
+		return 0, err
+	}
+
+	const fdSizeTag = "FDSize:"
+
+	for _, line := range strings.Split(string(status), "\n") {
+		if strings.HasPrefix(line, fdSizeTag) {
+			value := strings.TrimSpace(strings.TrimPrefix(line, fdSizeTag))
+			size, err := strconv.ParseUint(value, 10, 64)
+			if err != nil {
+				return 0, err
+			}
+			return size, nil
+		}
+	}
+
+	return 0, fmt.Errorf("tag %s not found in /proc/self/status", fdSizeTag)
+}
+
+func triggerEmfile() (*emfile, error) {
+	fdsize, err := getProcStatusFdSize()
+	if err != nil {
+		return nil, err
+	}
+
+	em := &emfile{}
+
+	if err := syscall.Getrlimit(syscall.RLIMIT_NOFILE, &em.limit); err != nil {
+		return nil, err
+	}
+
+	limit := em.limit
+	limit.Cur = fdsize
+
+	if err := syscall.Setrlimit(syscall.RLIMIT_NOFILE, &limit); err != nil {
+		return nil, err
+	}
+
+	for i := uint64(0); i < fdsize; i++ {
+		fd, err := syscall.Socket(syscall.AF_INET, syscall.SOCK_DGRAM, 0)
+		if err != nil {
+			return em, nil
+		}
+		em.fds = append(em.fds, fd)
+	}
+
+	return nil, fmt.Errorf("failed to trigger EMFILE")
+}
+
+func (em *emfile) undo() error {
+	if em == nil {
+		return nil
+	}
+
+	if err := syscall.Setrlimit(syscall.RLIMIT_NOFILE, &em.limit); err != nil {
+		return err
+	}
+	for _, fd := range em.fds {
+		syscall.Close(fd)
+	}
+
+	return nil
+}
```

Comment:
It would be good to have some description/comments for this code.

Reply:
Updated with comments about how we exhaust file descriptors.